### PR TITLE
Migrate `block`, `identity`, `module`, `contract`, `consensus`, `delegator` and `baker` client commands to GRPC V2.

### PR DESCRIPTION
**Pull request by vikt0r0**
## Purpose

Migrate the `block`, `identity`, `module`, `contract`, `consensus`, `delegator` and `baker` families of commands to use the GRPC V2 API; this is part of migrating the client to the GRPC V2 API feature.

Depends on:

https://github.com/Concordium/concordium-base/pull/317

https://github.com/Concordium/concordium-client/pull/215

## Changes
Implement helpers and make changes to glue together V2 API and commands of the client:

### Tasks
- Migrate `block`, `identity`, `module`, `contract`, `consensus`, `delegator` and `baker` commands.

### Client interface changes
- Slightly different (but hopefully improved) messages in case of API server errors and response conversion/invariant errors.

## Checklist
- [X] My code follows the style of this project.
- [X] The code compiles without warnings.
- [X] I have performed a self-review of the changes.
- [X] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: src/Concordium/Client/Runner.hs [hunk 1]
```diff
@@ -3346,15 +3439,15 @@ processDelegatorConfigureCmd baseCfgDir verbose backend txOpts cdCapital cdResta
...
-            Nothing -> [i||]
+      cooldownDate <- withClient backend $ do
+        bcpRes <- getBlockChainParametersV2 Best
+        case getResponseValue bcpRes of
+          Left (_, err) -> do
+            logError ["Could not get the delegator cooldown period: " <> err]
+            exitTransactionCancelled
+          Right v -> do
+            liftIO $ getDelegatorCooldown v
+      let cooldownString :: String = [i|The current baker cooldown would last until approximately #{cooldownDate}|]
```
Comment (vikt0r0) [inline src/Concordium/Client/Runner.hs:3450]: Should this say "delegator cooldown"?

File: src/Concordium/Client/Runner.hs [hunk 2]
```diff
@@ -1057,61 +1052,62 @@ getEncryptedAmountTransferData senderAddr ettReceiver ettAmount idx secretKey =
...
 getPoolStatusOrDie mbid = do
-  let (bid, passiveDelegation) = case mbid of
-        Nothing -> (0, True)
-        Just bakerId -> (bakerId, False)
-  infoValue <- logFatalOnError =<< withBestBlockHash Nothing (getPoolStatus bid passiveDelegation)
-  case AE.fromJSON $ grpcResponseVal infoValue of
-    AE.Error err -> logFatal ["Cannot decode pool status response from the node: " ++ err]
-    AE.Success Nothing -> logFatal ["Could not query pool status from the chain."]
-    AE.Success (Just ps) -> return ps
+  psRes <- getPoolInfoV2 Best (fromMaybe 0 mbid)
```
Comment (abizjak) [inline src/Concordium/Client/Runner.hs:1100]: I don't think this is correct. If `Nothing` is supplied you are meant to query passive delegation if I recall correctly.

File: src/Concordium/Client/Runner.hs [hunk 3]
```diff
@@ -1750,19 +1751,40 @@ processModuleCmd action baseCfgDir verbose backend =
...
-                               "the provided block does not exist:", Text.unpack bestBlock]
-        Just [] -> logInfo ["there are no modules in block " ++ Text.unpack bestBlock]
-        Just xs -> runPrinter $ printModuleList (bcModuleNameMap baseCfg) xs
+      -- Verify that the provided blockhash is valid.
+      case block of
+        Nothing -> return ()
+        Just b -> case readEither (Text.unpack b) of
+          Left _ -> logFatal ["Could not retrieve the list of modules",
+                              "the provided block hash is invalid, tried:",
+                              Text.unpack b]
```
Comment (abizjak) [inline src/Concordium/Client/Runner.hs:1760]: Won't this be on a separate line? Why?

File: src/Concordium/Client/Runner.hs [hunk 4]
```diff
@@ -1959,14 +2014,18 @@ processContractCmd action baseCfgDir verbose backend =
         Just (InvokerAccount nameOrAddr) -> Just . Types.AddressAccount . naAddr <$> getAccountAddressArg (bcAccountNameMap baseCfg) nameOrAddr
         Just InvokerContract{..} -> Just . Types.AddressContract . ncaAddr <$> getNamedContractAddress (bcContractNameMap baseCfg) icIndexOrName icSubindex
 
-      contrInfo <- withClient backend . withBestBlockHash block $ getContractInfo namedContrAddr
+      (bbHash, contrInfo) <- withClient backend $ do
+        bhInput <- readBlockHashOrDefault Best block
+        bHash <- extractResponseValueOrDie Queries.biBlockHash =<< getBlockInfoV2 bhInput
```
Comment (abizjak) [inline src/Concordium/Client/Runner.hs:2019]: So you are doing this a lot, but it is not necessary. It is not wrong, but each request returns in its header the block hash that was used, so if you supply Best you will get back a header blockHash with the actual hash that was used.

this would save these requests. I don't want you to spend time changing things now, but wanted to bring it up.

File: src/Concordium/Client/Runner.hs [hunk 5]
```diff
@@ -1978,34 +2037,40 @@ processContractCmd action baseCfgDir verbose backend =
...
-              let eventsMsg = case mapMaybe (fmap (("  - " <>) . Text.pack) . showEvent verbose Nothing) rcrEvents of
-                                [] -> Text.empty
-                                evts -> [i|- Events:\n#{Text.intercalate "\n" evts}|]
-              returnValueMsg <- mkReturnValueMsg rcrReturnValue schemaFile modSchema contractName updatedReceiveName False
-              logSuccess [[iii|Invocation resulted in success:\n
-                               - Energy used: #{showNrg rcrUsedEnergy}
-                               #{returnValueMsg}
-                               #{eventsMsg}
-                               |]]
+      -- VH/FIXME: Should the blockhash input in the following be `Best` or `Given bbHash`?
```
Comment (abizjak) [inline src/Concordium/Client/Runner.hs:2041]: Yes, it should be the given block. That's the most consistent, even if it was different before.

File: src/Concordium/Client/Runner.hs [hunk 6]
```diff
@@ -3111,12 +3214,14 @@ processBakerUpdateStakeBeforeP4Cmd baseCfgDir verbose backend txOpts ubsStake =
...
-          logError ["Could not reach the node to get the baker cooldown period."]
-          exitTransactionCancelled
+      cooldownDate <- withClient backend $ do
+        bcpRes <- getBlockChainParametersV2 Best
+        case getResponseValue bcpRes of
+          Left (_, err) -> do
+            logError ["Could not get the baker cooldown period: " <> err]
+            exitTransactionCancelled
+          Right v -> do
+            getBakerCooldown v
```
Comment (abizjak) [inline src/Concordium/Client/Runner.hs:3224]:
```suggestion
          Right v -> getBakerCooldown v
```

File: src/Concordium/Client/Runner.hs (hunk 1 above)
Comment (abizjak) [reply src/Concordium/Client/Runner.hs:3462]: Yes.

File: src/Concordium/Client/Runner.hs [hunk 7]
```diff
@@ -3498,31 +3605,31 @@ processLegacyCmd action backend =
...
-        printResponseValueAsJSON
+          getBlockPendingUpdatesV2 >>=
+            printResponseValueAsJSON
     GetBlockSpecialEvents block ->
       withClient backend $
         readBlockHashOrDefault Best block >>=
-        getBlockSpecialEventsV2 >>=
-        printResponseValueAsJSON
+          getBlockSpecialEventsV2 >>=
+            printResponseValueAsJSON
```
Comment (abizjak) [inline src/Concordium/Client/Runner.hs:3622]: Why were these changes necessary? Please avoid noise.

File: src/Concordium/Client/Runner.hs [hunk 8]
```diff
@@ -3951,18 +4058,23 @@ processCredential ::
   => BSL.ByteString
   -> Int
   -> ClientMonad m Types.BareBlockItem
-processCredential source networkId =
+processCredential source _networkId =
```
Comment (abizjak) [inline src/Concordium/Client/Runner.hs:4061]: Why is the parameter not getting removed alltogether?

**Review by abizjak — APPROVED** [review]
Please address the issues. Otherwise looks ok.

File: src/Concordium/Client/Runner.hs (hunk 2 above)
Comment (vikt0r0) [reply src/Concordium/Client/Runner.hs:1100]: Thanks for pointing that out. I will have to look into that a bit further.

File: src/Concordium/Client/Runner.hs (hunk 3 above)
Comment (vikt0r0) [reply src/Concordium/Client/Runner.hs:1760]: Yes, I do not know either. Probably it should just be a single line like everything else?

File: src/Concordium/Client/Runner.hs (hunk 4 above)
Comment (vikt0r0) [reply src/Concordium/Client/Runner.hs:2019]: Neat. I'll change it.

File: src/Concordium/Client/Runner.hs (hunk 5 above)
Comment (vikt0r0) [reply src/Concordium/Client/Runner.hs:2041]: 👍🏼

File: src/Concordium/Client/Runner.hs (hunk 7 above)
Comment (vikt0r0) [reply src/Concordium/Client/Runner.hs:3622]: They were not strictly, but I did not know how sequencing should be formatted, and this was consistent with how I ended up doing things elsewhere.

Comment (vikt0r0) [reply src/Concordium/Client/Runner.hs:3622]: Do I revert this now?

File: src/Concordium/Client/Runner.hs (hunk 8 above)
Comment (vikt0r0) [reply src/Concordium/Client/Runner.hs:4061]: It wa tempting to do, but I didn't know if this was for a separate PR, so I was a bit wary.

I'll just do it here then?

I do not remember this is the case, or if I perhaps already removed it, but should it be removed from the CLI if it is  there?

File: src/Concordium/Client/Runner.hs (hunk 4 above)
Comment (abizjak) [reply src/Concordium/Client/Runner.hs:2019]: As I wrote, if this is going to delay merging this PR then do not do it.

File: src/Concordium/Client/Runner.hs (hunk 3 above)
Comment (abizjak) [reply src/Concordium/Client/Runner.hs:1760]: Yes.

File: src/Concordium/Client/Runner.hs (hunk 8 above)
Comment (abizjak) [reply src/Concordium/Client/Runner.hs:4061]: We just used default 100 I think. This parameter was never used even in the old API.

File: src/Concordium/Client/Runner.hs (hunk 7 above)
Comment (vikt0r0) [reply src/Concordium/Client/Runner.hs:3622]: fourmolu is not used in the client is it? I recall using it on a file in the client and it pretty much changed up everything.

Is what you are proposing that I use fourmolu config of concordium-base to see how it formats things (in this case sequencing) and then go from that that as a reference?

File: src/Concordium/Client/Runner.hs (hunk 4 above)
Comment (vikt0r0) [reply src/Concordium/Client/Runner.hs:2019]: I don't think it will, but if it seems to be cumbersome I'll leave it as is.

Comment (vikt0r0) [reply src/Concordium/Client/Runner.hs:2019]: I gave it a shot, but it took too long, so I opted not to.
